Features/abi 1/cart count into gtm

Abdeve opened the pull request:
Comments added into `docs/notes-on-new-features.md`!

AndreiBoris committed:
ABI-1: Add cart schema and query resolver with empty data

AndreiBoris committed:
ABI-1: Show simple cart products in sidebar

No interaction is available at this time.

AndreiBoris committed:
ABI-1: Simple add to cart function

No validation.
No ability to remove cart items.
No persistence beyond current run time of the application.
No association between the cart and the user. A logged in user is assumed.

AndreiBoris committed:
ABI-1: Inform GTM about number of items in cart through data layer

AndreiBoris committed:
ABI-1: Documentation on the current implementation

AndreiBoris committed:
ABI-1: Fix missing definitions

A result of having moved from one repository to another in a different github account.

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+  return cartProducts ? cartProducts.reduce((total, cartProduct) => {
+    return total += cartProduct.quantity;
+  }, 0) : 0;
+}
```
This looks like it breaks 2 rules:

- See 17.2 Selection operators in place of control statements https://airbnb.io/javascript/ 
- Unnecessary complexity (return statements). We like [to follow ESLint Complexity rule](https://eslint.org/docs/rules/complexity). so the extra return statements will break this rule more easily
- This function should assume you have cartProducts, and you should check before passing it into this function
- I would also rename the function to do with quantity, since product-count can mean # of unique products, not necessarily the quantity. Imagine a cart with 2 products, but 3 quantity each. It sounds like your function would return 2, instead of 6.

Suggestion:
```
const getCartQuantityTotal = cartProducts => 
cartProducts.reduce((total, cartProduct) => total += cartProduct.quantity,0)
```

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+      return (
+        <p>
+          Cart is empty
+        </p>
```
I would do an early return if validation of cartProducts fails. So, something like this

```
const component = (data) => {

	const isValid = validate(data);
	
	if (!isValid) {
	  return (
	    <p>
	      Cart is empty
	    </p>
	  )
	}
	
	getCount(..)
		
	return <valid>

}

```

It simplifies the logic of the components and return statements

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+class Cart extends Component {
+  showCartProducts() {
+    const { cartProducts, cartLoading } = this.props;
+
```
In this codebase, this is where I would verify that cartProducts exists and is valid, so I wouldn't have to verify in the subsequent if-statements or the getCount function.

Ideally, you would do it as close a possible to receiving it from the data layer (graphql, not GTM dataLayer).
Otherwise, you end up with overly defensive code and very verbose, constantly checking on each layer if the data is valid.

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+import CartItem from './CartItem';
+import GTMConstants from '../helpers/GTMConstants'
+
+const GET_CART_PRODUCTS = gql`
```
you can add the export here, instead of at the bottom of the file

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+          Cart
+        </h1>
+        {this.showCartProducts()}
+      </div>
```
Consider `<Fragments>` since the div doesn't do anything, it'll reduce the nesting

omerose commented on server/modules/merchant/resolvers.js:
```diff
@@ -1,8 +1,35 @@
+      }
+
+      const newCartProduct = {
+        cartProductId: v4(),
```
This is a very bizarre choice.

The products have IDs already and on line 18 you check productId not cartProductId.

I see the only use of this variable is for the key={} prop inside the `<CartItem>`. However, you should be constructing the key out of things you have locally. You can combine name + id, to get a very unique key. The react spec is key={} just needs to be unique on the front-end between all the components, the likelihood of a conflict is very low, even if you just use ID alone.

omerose replied:
Perhaps there's some caching magic, I'm not aware of in Apollo/graphql, so I'll concede to that. However, I would still construct the cartProductId with the existing ID + name, so we're not introducing an entire library to the codebase just to generate IDs

Abdeve replied:
That's fair! The decision was made to reflect what would normally identify CartProduct records which are separate from Product records in a couple of ways:

1. Multiple users might have each have a given Product in their cart and as such a new identifier is needed to identify records such that there may be many of the same ProductId. A composite key between productId and UserId might also be used in this case but that's not something I am familiar with in the context of GraphQL.
2. To account for the possibility that Products may later be edited by admins and to preserve the CartProduct if it made it into an as-of-now undefined Order record.

However, given that these are massive assumptions that I didn't lay out, your critique makes total sense to me!

omerose commented on docs/notes-on-new-features.md:
```diff
@@ -0,0 +1,9 @@
+
+This change assumed a logged in user and valid data being passed from React to the Apollo server. Additionally, the implementation is such that the database pulled from mockMerchantData but is then stored in memory before being requeried using `GET_CART_PRODUCTS`.
+
+Whenever an item is added to the cart in this way, I use the `dataLayer` on Google Tag Manager to update the `itemsInCartCount` variable which can be used by the marketing team to some effect to target users who appear inclined to make a purchase.
```
There are no tracking-events connected to the itemInCartCount. So this feature seems incomplete.

omerose replied:
It would be important to see how you think this event should be tracked, since some implementations would be incorrect, and others more correct

omerose commented on client/src/App.js:
```diff
@@ -1,13 +1,33 @@
+              <Cart />
+            </Col>
+          </Row>
+        </Container>
```
Good use. Although today we prefer Flex-box/Grid, but 2019 was a long time ago (when this test was created)

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+          {
+            cartProducts.map(cartProduct => {
+              const { cartProductId, name, quantity, price } = cartProduct
+              return (
```
Could've been shorter by de-structuring in the input variable

omerose commented on client/src/helpers/GTMConstants.js:
```diff
@@ -0,0 +1,4 @@
+export default {
+  GTM_CONTAINER_ID: 'GTM-K5R56QB',
+  DATA_LAYER_CART_ITEM_COUNT: 'itemsInCartCount',
+};
```
good use of constants instead of using bare-words everywhere

omerose commented on client/src/components/Products.js:
```diff
@@ -54,7 +74,29 @@ class ProductsList extends Component {
+                        onClick={addToCart}
+                      >
+                        {loading ? 'Adding to Cart' : 'Buy'}
+                      </Button>
```
clean file, looks good

omerose commented on server/modules/merchant/resolvers.js:
```diff
@@ -1,8 +1,35 @@
+        quantity: 1,
+        ...args,
+      };
+      cart.push(newCartProduct);
```
I don't like use of mutative functions like push, splice, etc.

It's also deceiving that you have 'const cart = []', yet you are mutating it. If you have to do it this way, let's use:

```
let cart = []

... later
cart = [ ...cart, newCartProduct ]

```

omerose commented on client/src/components/Cart.js:
```diff
@@ -0,0 +1,88 @@
+      }}
+    </Query>
+  );
+};
```
Some places you have the redundant return, other places you don't. It's a style choice, but stick to one to be consistent.

Haven't tested, but something like this?
```
const withCartProducts = Component => props => (
    <Query query={GET_CART_PRODUCTS}>
      {({ loading, data }) => (
          <Component cartLoading={loading} cartProducts={data && data.cart} {...props} />
        );
      }
    </Query>
  );
```

omerose commented on client/src/styles.css:
```diff
@@ -1,5 +1,4 @@
 .product-card {
-    width: 50%;
```
I would've liked to see something for mobile, since your page doesn't fit on mobile

omerose commented on docs/notes-on-new-features.md:
```diff
@@ -0,0 +1,9 @@
+# Tracking number of items in cart
+
+The feature that I've implemented in a simple cart. By clicking the Buy button on the products listed below.
```
```suggestion
The feature that I've implemented is a simple cart. By clicking the Buy button on the products listed below.
```

omerose commented on docs/notes-on-new-features.md:
```diff
@@ -0,0 +1,9 @@
+
+![Shop Page Default Look and browse](add-to-cart.gif)
+
+This change assumed a logged in user and valid data being passed from React to the Apollo server. Additionally, the implementation is such that the database pulled from mockMerchantData but is then stored in memory before being requeried using `GET_CART_PRODUCTS`.
```
```suggestion
This change assumed a logged in user and valid data being passed from React to the Apollo server. Additionally, the implementation is such that the database is pulled from mockMerchantData but is then stored in memory before being requeried using `GET_CART_PRODUCTS`.
```

omerose reviewed (changes requested):
It's ok for the code that's here. However it's lacking from a complete test/solution expected. 

I thought you were going to see our grading rubric and therefore would see the additional requirements of this test. So it seems that you received partial instructions. I apologize that the instructions that were sent to you were incomplete.

You can continue this test for a complete solution based on the instructions (you are like 70% of the way there), or you can leave it as-is, correct your mistakes, and I will factor into account that you received partial instructions.

Here's the full instructions we sent to every candidate.

```
The react-interview-test is open-ended for a fullstack role, so the instructions say implement ANY feature. 
Please disregard that and use any existing functionality, or feel free to add a very small feature for this next step. You will be graded according to the tracking/plan, not any features you implement (so don't worry about making them pretty).
Create 2 tracking events (scroll, pageview, button-clicks, etc). Please select a meaningful thing to track and explain why that was selected.
Additionally, create a plan (documentation) to track the conversion funnel of this product (please make assumptions on what the completed product will be like) and include the 2 tracking events you created.
Submission requirements:
 - At least 2 tracking events
 - Conversion funnel plan
 - Instructions to a manager on which events were tracked and where to access them.
```

Cheers,

Abdeve commented:
Thank you for leaving the comments and providing the additional instruction, Omer! I'll take a closer look at these at this evening and re-submit tonight with the corrections along with new work following the additional instructions. Cheers!

AndreiBoris committed:
ABI-1: Fix code style errors with cart quantity function

AndreiBoris committed:
ABI-1: Change count to quantity for GTM dataLayer variable

Better reflects the data that this variable holds and keeps it consistent with the renamed functions from commit 8c76d2de32f4339d54a387d1a5ab87d889379d8b

AndreiBoris committed:
ABI-1: Breakdown Cart to include CartContents component

Cart is responsible for fetching data and CartContents is responsible for working with that data.

Remove unnecessary DOM element nesting in Cart.js

AndreiBoris committed:
ABI-1: Move export of GET_CART_PRODUCTS next to its definition

AndreiBoris committed:
ABI-1: Get rid of the unneeded cartProductId on CartProduct schema

AndreiBoris committed:
ABI-1: Keep redundant returns in all arrow functions to keep consistent

Initial codebase used these so let's not change it.

AndreiBoris committed:
ABI-1: Basic responsiveness to added to css

Abdeve commented:
Hi! Just a heads up: I've added the changes to most of the comments you made earlier this evening but have had to deal with something urgent so haven't had a chance to look at the proper tracking assignment. I'll review and give the actual assignment a shot tomorrow evening, if that can work!

omerose commented:
Regarding the tracking assignment, that's fine

omerose commented on client/src/App.js:
```diff
@@ -18,16 +17,10 @@ class App extends Component {
-            </Col>
-          </Row>
-        </Container>
+        <main className="react-interview-container">
+          <Cart />
+          <Products />
+        </main>
```
Looks much cleaner, good work

omerose commented on client/src/components/CartContents.js:
```diff
@@ -0,0 +1,61 @@
+    },
+  });
+  
+  if (isCartEmpty(props)) {
+    return <p>
+      Cart is empty
+    </p>;
+  }
+
+  return (
+    <ListGroup>
+      {
+        cartProducts.map(({ productId, name, quantity, price }) => {
+          return (
+            <CartItem 
+              key={productId}
+              name={name}
+              quantity={quantity}
+              price={price}
+            />
+          );
+        })
+      }
+    </ListGroup>
+  );
+}
```
Awesome, much easier to follow!

omerose reviewed (commented):
Better. Looking forward to the documentation and tracking

AndreiBoris committed:
ABI-1: Add placeholder checkout button with gtmclickid

AndreiBoris committed:
ABI-1: Add documentation of tracking plan

Abdeve commented:
Hi @omerose ! I've pushed up changes with some changes to the tracking events and documentation regarding the tracking plan.

I didn't touch the `cart` just to focus on the tracking portion of the assignment.

omerose commented on docs/tracking-plan.md:
```diff
@@ -0,0 +1,63 @@
+# Full stack (data) submission documentation
+
+## Modification to the app
+
+The feature that I've implemented is a simple cart. By clicking the Buy button on the products listed below.
+
+![Add to cart](add-to-cart.gif)
+
+This change assumed a logged in user and valid data being passed from React to the Apollo server. Additionally, the implementation is such that the "database" is stored in memory inside the `cart` array which is then returned by `Query cart` and is then mutated by `Mutation addToCart`. In a proper implementation this data would be stored in the database but this is too tedious to implement in thi ...
+
+We have also added a "Checkout" button to the cart. This button doesn't do anything but is connected to our GTM container tracking. See the Tracking Plan below for more details.
```
Clear documentation thank you. I have reviewed it fully. We will be contact.

Abdeve replied:
Sounds great - looking forward to hearing from you and Bonsai. Have a great evening!
